kernel/src/debug.rs

```diff
@@ -271,118 +264,50 @@ impl DebugWriter {
         self.count.get()
     }
 
-    /// Convenience method that writes (end-start) bytes from bytes into the
-    /// internal debug buffer.
-    fn write_buffer(&self, start: usize, end: usize, bytes: &[u8]) {
-        if end < start {
-            panic!("wb bounds: start {} end {} bytes {:?}", start, end, bytes);
-        }
-        self.internal_buffer.map(|in_buffer| {
-            for (dst, src) in in_buffer[start..end].iter_mut().zip(bytes.iter()) {
-                *dst = *src;
-            }
-        });
-    }
-
     /// Write as many of the bytes from the internal_buffer to the output
     /// mechanism as possible.
     fn publish_str(&self) {
         // Can only publish if we have the output_buffer. If we don't that is
         // fine, we will do it when the transmit done callback happens.
         self.output_buffer.take().map(|out_buffer| {
-            let head = self.head.get();
-            let tail = self.tail.get();
-            let len = self
-                .internal_buffer
-                .map_or(0, |internal_buffer| internal_buffer.len());
-
-            // Want to write everything from tail inclusive to head
-            // exclusive
-            let (start, end) = if tail > head {
-                // Need to pass subscribe a contiguous buffer, so first
-                // write from tail to end of buffer. The completion
-                // callback will see that the buffer's not empty and
-                // call again to write the rest (tail will be 0)
-                let start = tail;
-                let end = len;
-                (start, end)
-            } else if tail < head {
-                let start = tail;
-                let end = head;
-                (start, end)
-            } else {
-                panic!("Consistency error: publish empty buffer?")
-            };
-
-            // Check that we aren't writing a segment larger than the output
-            // buffer.
-            let real_end = start + cmp::min(end - start, out_buffer.len());
-
-            self.internal_buffer.map(|internal_buffer| {
-                for (dst, src) in out_buffer
-                    .iter_mut()
-                    .zip(internal_buffer[start..real_end].iter())
-                {
-                    *dst = *src;
+            let mut count = 0;
+
+            let ring_buffer = self.internal_buffer.take().unwrap();
```

Comment:
`unwrap` always makes me nervous, but I'm not sure what the behavior here should be if the buffer isn't available?

Reply:
Good point! I guess I should use `.map()` instead of `.take()`. Will try that & fix the other comments.